File: util/src/main/java/jcog/pri/UnitPrioritizable.java
package jcog.pri;

import jcog.util.FloatFloatToFloatFunction;

/** marker interface for implementations which limit values to the range of 0..1.0 (and deleted, NaN) */
public interface UnitPrioritizable extends Prioritizable {

    /**
     * assumes 1 max value (Plink not NLink)
     */
    default float priAddOverflow(float inc /* float upperLimit=1 */) {

        if (inc <= EPSILON)
            return (float) 0;

        float[] beforeAfter = priDelta(new FloatFloatToFloatFunction() {
            @Override
            public float apply(float x, float y) {
                return ((x != x) ? (float) 0 : x) + y;
            }
        }, inc);

        float after = beforeAfter[1];
        float before = beforeAfter[0];
        float delta = (before != before) ? after : (after - before);
        return Math.max((float) 0, inc - delta); //should be >= 0
    }
}

File: util/src/main/java/jcog/pri/UnitPrioritizable.kt
package jcog.pri

import jcog.util.FloatFloatToFloatFunction

/** marker interface for implementations which limit values to the range of 0..1.0 (and deleted, NaN)  */
interface UnitPrioritizable : Prioritizable {

    /**
     * assumes 1 max value (Plink not NLink)
     */
    @kotlin.jvm.JvmDefault
    fun priAddOverflow(inc: Float /* float upperLimit=1 */): Float {

        if (inc <= jcog.pri.ScalarValue.EPSILON)
            return 0.toFloat()

        val beforeAfter = priDelta(FloatFloatToFloatFunction { x, y -> (if (x != x) 0.toFloat() else x) + y }, inc)

        val after = beforeAfter[1]
        val before = beforeAfter[0]
        val delta = if (before != before) after else after - before
        return Math.max(0.toFloat(), inc - delta) //should be >= 0
    }
}
